src/query/storages/common/table-meta/src/meta/v3/segment.rs

```diff
@@ -0,0 +1,132 @@
+//  Copyright 2021 Datafuse Labs.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+use std::io::Error;
+use std::io::ErrorKind;
+use std::io::Write;
+use std::sync::Arc;
+
+use common_exception::Result;
+use common_expression::TableField;
+use serde::Deserialize;
+use serde::Serialize;
+use zstd::Encoder;
+
+use crate::meta::statistics::FormatVersion;
+use crate::meta::v2::BlockMeta;
+use crate::meta::Statistics;
+use crate::meta::Versioned;
+
+/// A segment comprises one or more blocks
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct SegmentInfo {
+    /// format version
+    format_version: FormatVersion,
+    /// blocks belong to this segment
+    pub blocks: Vec<Arc<BlockMeta>>,
+    /// summary statistics
+    pub summary: Statistics,
+}
+
+impl SegmentInfo {
+    #[inline]
+    pub fn version(&self) -> FormatVersion {
+        self.format_version
+    }
+}
+use super::super::v0;
+use super::super::v1;
+use super::super::v2;
+
+impl SegmentInfo {
+    pub fn from_v0(s: v0::SegmentInfo, fields: &[TableField]) -> Self {
+        let summary = Statistics::from_v0(s.summary, fields);
+        Self {
+            format_version: SegmentInfo::VERSION,
+            blocks: s
+                .blocks
+                .into_iter()
+                .map(|b| Arc::new(BlockMeta::from_v0(&b, fields)))
+                .collect::<_>(),
+            summary,
+        }
+    }
+
+    pub fn from_v1(s: v1::SegmentInfo, fields: &[TableField]) -> Self {
+        let summary = Statistics::from_v0(s.summary, fields);
+        Self {
+            format_version: SegmentInfo::VERSION,
+            blocks: s
+                .blocks
+                .into_iter()
+                .map(|b| Arc::new(BlockMeta::from_v1(b.as_ref(), fields)))
+                .collect::<_>(),
+            summary,
+        }
+    }
+
+    pub fn from_v2(s: v2::SegmentInfo) -> Self {
+        Self {
+            format_version: SegmentInfo::VERSION,
+            blocks: s.blocks,
+            summary: s.summary,
+        }
+    }
+
+    pub fn to_bytes(&self) -> Result<Vec<u8>> {
```

Comment:
Why not using bincode's serde support directly?

Reply:
https://docs.rs/bincode/2.0.0-rc.3/bincode/serde/index.html

Reply:
We want to support partial read functionality, such as only reading blocks, so we have designed a separate file format that includes header information: version, format (binary, messagepack..), compress (zstd, snap..), blocks size, summary size.

Reply:
should add comments on this method.

Reply:
> Since segmengt is so small after compression, it's better to read all content in memory instead read partial data from underlying storage.

Next, I will complete many comments.

Reply:
To avoid the deserialization cost from memory, we can try https://github.com/rkyv/rkyv

Reply:
Another purpose of the header is to store encoding formats(bincode, messagepack, json...) and compression methods(zstd, compress, none...), so that in the future, without upgrading the version, the storage format and compression method can be seamlessly replaced, making it more scalable.

Reply:
@Xuanwo  It's already been discussed with @dantengsky , he needs partial deserialization.